dali/daliadmin/daliadmin.cpp

```diff
@@ -2794,114 +2798,183 @@ static void wuidCompress(const char *match, const char *type, bool compress)
+            const char *w = wuid.str();
+            associatedWithWuid = nullptr != wuids.find(*w);
+            if (!associatedWithWuid)
             {
-                StringBuffer wuid;
-                RE.substitute(wuid,"#1");
-                const char *w = wuid.str();
-                bool found = NULL != wuids.find(*w);
-                const char *uid = gd.queryProp("@uid");
-                if (!found)
+                generateDllsToDelete.push_back(&gd);
+                entryRemoved = true;
+            }
+        }
+
+        if (!entryRemoved && removeStaleLocations)
+        {
+            std::vector<IPropertyTree *> locationsToRemove;
+            Owned<IPropertyTreeIterator> locationIter = gd.getElements("location");
+            ForEach(*locationIter)
+            {
+                IPropertyTree &location = locationIter->query();
+                const char *host = location.queryProp("@ip");
+                if (!isEmptyString(host))
                 {
-                    uidsToDelete.append(uid);
-                    indexToDelete.append(index);
+                    IpAddress ip(host);
+                    StringBuffer ipStr;
+                    ip.getIpText(ipStr);
+                    const auto &it = machineMap.find(ipStr.str());
+                    if (it == machineMap.end()) // not found, remove it.
+                    {
+                        if (verbose)
+                            PROGLOG("Removing stale for %s at location %s", name, host);
+                        locationsToRemove.push_back(&location);
+                        ++removedLocations;
+                    }
                 }
             }
+            if (fix)
+            {
+                for (auto &location: locationsToRemove)
+                    gd.removeTree(location);
+            }
         }
-        ++index;
     }
-    PROGLOG("%d out of %d workunit files not associated with any workunit. Took %d ms", indexToDelete.ordinality(), index, ts.elapsed());
-    ts.reset(0);
+    PROGLOG("%u out of %u generated dll entries not associated with any workunit, and %u stale locations found. Took %u ms", (unsigned)generateDllsToDelete.size(), totalCount, removedLocations, ts.elapsed());
 
+    ts.reset(0);
     IArrayOf<IDllEntry> removedEntries;
     unsigned numDeleted = 0;
-    ForEachItemInRev(d, indexToDelete)
-    {
-        const char *uid = uidsToDelete.item(d);
-        unsigned index = indexToDelete.item(d);
-        StringBuffer path("GeneratedDll[");
-        path.append(index).append("]");
-        IPropertyTree *gd = root->queryPropTree(path.str());
-        if (NULL == gd)
-            throwUnexpected();
-        const char *uidQuery = gd->queryProp("@uid");
-        if (0 != strcmp(uid, uidQuery))
-            throw MakeStringException(0, "Expecting uid=%s @ GeneratedDll[%d], but found uid=%s", uid, index, uidQuery);
+    for (auto &toDelete: generateDllsToDelete)
+    {
         if (verbose)
```

Comment:
No, I still want verbose=true to trace when fix=false.
It could alternate the tracing to "would remove", but last line of report will say '... entries to remove ... use 'fix=true'